packages/ethereum/contracts/HoprDistributor.sol

```diff
@@ -0,0 +1,264 @@
+     * @param _token The token which we will mint
+     */
+    constructor(HoprToken _token, uint32 _startTime, uint128 _maxMintAmount) public {
+        startTime = _startTime;
+        token = _token;
+        maxMintAmount = _maxMintAmount;
+    }
+
+    /**
+     * @dev Revokes the ability for an account to claim on the
+     * specified schedule.
+     * @param account the account to crevoke
+     * @param scheduleName the schedule name
+     */
+    function revokeAccount(
+        address account,
+        string calldata scheduleName
+    ) external onlyOwner {
+        require(allocations[account][scheduleName].amount != 0, "Allocation must exist");
+        allocations[account][scheduleName].revoked = true;
+    }
+
+    /**
+     * @dev Adds a schedule, the schedule must not already exist.
+     * Owner is expected to insert values in ascending order,
+     * each element in arrays {durations} and {percents} is meant to be
+     * related.
+     * @param durations the durations for each schedule period in seconds (6 months, 1 year)
+     * @param percents the percent of how much can be allocated during that period,
+     * instead of using 100 we scale the value up to {MULTIPLIER} so we can have more accurate
+     * "percentages".
+     */
+    function addSchedule(
+        uint32[] calldata durations,
+        uint32[] calldata percents,
+        string calldata name
+    ) external onlyOwner {
+        require(schedules[name].durations.length == 0, "Schedule must not exist");
+        require(durations.length == percents.length, "Durations and percents must have equal length");
+
+        uint32 lastDuration = 0;
+        for (uint256 i = 0; i < durations.length; i++) {
+            require(lastDuration <= durations[i], "Durations must be added in ascending order");
+            lastDuration = durations[i];
+            require(percents[i] <= MULTIPLIER, "Percent provided must be smaller or equal to MULTIPLIER");
+        }
+
+        schedules[name] = Schedule(durations, percents);
+
+        emit ScheduleAdded(durations, percents, name);
+    }
+
+    /**
+     * @dev Adds allocations, all allocations will use the schedule specified,
+     * schedule must be created before and account must not have an allocation
+     * in the specific schedule.
+     * @param accounts accounts to create allocations for
+     * @param amounts total amount to be allocated
+     * @param scheduleName the schedule name
+     */
+    function addAllocations(
+        address[] calldata accounts,
+        uint128[] calldata amounts,
+        string calldata scheduleName
+    ) external onlyOwner {
+        require(schedules[scheduleName].durations.length != 0, "Schedule must exist");
+        require(accounts.length == amounts.length, "Accounts and amounts must have equal length");
+
+        for (uint256 i = 0; i < accounts.length; i++) {
+            require(allocations[accounts[i]][scheduleName].amount == 0, "Allocation must not exist");
```

Comment:
Here we should also check that enough tokens are actually available to be minted.
So increase a global `totalToBeMinted +=  allocations[accounts[i]][scheduleName].amount;` and then check `require(totalToBeMinted < maxMintAmount)`